### PR TITLE
super: fix GCC linkage issue in Linux

### PR DESCRIPTION
**Pull request by danielnachun**
- [ ] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [ ] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [ ] Have you successfully run `brew style` with your changes locally?
- [ ] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

### CODE REVIEW

**Comment by danielnachun** [comment]
This is my second attempt to fix the issues relating to GCC linker flag ordering in the superenv in Linux.  To recap why we need this:

1) For users without a host GCC at least version 5, we have to place symlinks for the GCC 5 runtime libs in `$HOMEBREW_PREFIX/lib`, because we don't add RPATHs for brewed GCC 5 to our binaries built in CI.  On systems with a new enough host GCC this doesn't matter because we can use the host GCC runtime libs, but on systems that need brewed GCC 5, we have to do this so that bottles built in CI can find the brewed GCC 5 runtime libs.  We could add the RPATH for brewed GCC 5 when pouring bottles using `patchelf.rb`, but that is more complex fix and would require users to repour all bottles with GCC 5 runtime dependencies.

2) The superenv adds `-L$HOMEBREW_PREFIX/lib` to the LDFLAGS when building from source.  My previous attempt in https://github.com/Homebrew/brew/pull/10606 to fix this created breakage because I tried to remove `-L$HOMEBREW_PREFIX/lib`, which it turns out we still need for some things.  

3) We have a custom specs for newer GCC versions which adds the proper linker flags and RPATHS to all binaries built by that newer GCC.   However, because of point 2 above, the superenv adds `-L$HOMEBREW_PREFIX/lib` **before** the linker flags added by the GCC specs file.  And because of point 1 above, this means the linker finds the GCC runtime libs for GCC 5 **before** finding the runtime libs for newer GCC.  This is possible because the sonames for GCC runtime libs are not always incremented between major versions even though the ABI is not the same (this is because the ABI is backwards compatible but not forwards compatible).  

This PR implements a solution suggested by @Bo98 to add some additional logic to check if the compiler being used is GCC, and if so, to add the LDFLAG for the newer GCC **before** `-L$HOMEBREW_PREFIX/lib`.  No flags are removed, and the LDFLAG we are adding is already added by the specs file.  This essentially creating an extra copy of that flag and just making sure it comes before any other LDFLAGS.  

I have tested this on a system with brewed GCC 5 while building a package with a newer brewed GCC, and it works as expected.  Let me know if this needs further explanation as this is a particularly arcane problem!

File: Library/Homebrew/extend/ENV/super.rb [hunk 1]
```diff
@@ -199,10 +199,15 @@ def homebrew_extra_library_paths
 
   sig { returns(T.nilable(PATH)) }
   def determine_library_paths
-    paths = [
-      keg_only_deps.map(&:opt_lib),
-      HOMEBREW_PREFIX/"lib",
-    ]
+    paths = []
+    if compiler.match?(GNU_GCC_REGEXP)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/extend/ENV/super.rb:203]:
```suggestion
   # some comment explaining why this is done
    if compiler.match?(GNU_GCC_REGEXP)
```

File: Library/Homebrew/extend/ENV/super.rb [hunk 2]
```diff
@@ -199,10 +199,15 @@ def homebrew_extra_library_paths
...
+      f = gcc_version_formula(compiler.to_s)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/extend/ENV/super.rb:204]: Need to catch the `FormulaUnavailableError` here in case it doesn't exist.

**Review by MikeMcQuaid — COMMENTED** [review]
Looks good so far!

File: Library/Homebrew/extend/ENV/super.rb [hunk 3]
```diff
@@ -199,10 +199,15 @@ def homebrew_extra_library_paths
...
+      paths << f.opt_lib/"gcc/#{f.version.major}" if f.any_version_installed?
```
Comment (carlocab) [inline Library/Homebrew/extend/ENV/super.rb:205]: Minor style nit.
```suggestion
      paths << f.opt_lib/"gcc"/f.version.major if f.any_version_installed?
```
We don't need interpolation here. (Probably doesn't make a huge difference, but I find this more readable.)

**Review by MikeMcQuaid — APPROVED** [review]

**Pull request merged**